temporal101/Exercise3.hs

```diff
@@ -0,0 +1,143 @@
+module Main where
+
+import Control.Monad.IO.Class (liftIO)
+import Control.Monad.Logger (defaultOutput, runStdoutLoggingT)
+import Control.Monad.Trans.Reader (runReaderT)
+import Data.Aeson (FromJSON, ToJSON)
+import Data.ByteString.Char8 qualified as BS
+import Data.Text (pack, unpack, Text)
+import Data.UUID qualified as UUID
+import Data.UUID.V4 qualified as UUID.V4
+import DiscoverInstances (discoverInstances)
+import GHC.Generics (Generic)
+import Network.HTTP.Simple (httpBS, getResponseBody, parseRequest)
+import RequireCallStack (RequireCallStack, provideCallStack)
+import System.IO (stdout)
+import Temporal.Activity (Activity)
+import Temporal.Client (mkWorkflowClientConfig, workflowClient)
+import Temporal.Client qualified as Client
+import Temporal.Core.Client (connectClient, defaultClientConfig)
+import Temporal.Duration (seconds)
+import Temporal.Payload (JSON (JSON))
+import Temporal.Runtime (TelemetryOptions (..), initializeRuntime)
+import Temporal.TH (WorkflowFn, ActivityFn)
+import Temporal.TH qualified
+import Temporal.Worker (Worker, WorkerConfig, startWorker)
+import Temporal.Worker qualified as Worker
+import Temporal.Workflow (Workflow, WorkflowId (..))
+import Temporal.Workflow qualified as Workflow
+import UnliftIO.Exception (bracket)
+
+data ActivityInput = ActivityInput
+  { name :: Text
+  }
+  deriving stock (Generic, Show)
+  deriving anyclass (FromJSON, ToJSON)
+
+-- | An Activity that calls a local webserver. Activity has a MonadIO
+-- instance so we don't need to provide anything in the environment,
+-- which remains @()@.
+getSpanishGreeting :: ActivityInput -> Activity () Text
+getSpanishGreeting input = do
+  req <- parseRequest . unpack $ "http://localhost:9001/hello/" <> input.name
+  resp <- httpBS req
+  pure . pack . BS.unpack $ getResponseBody resp
```

Comment:
```suggestion
  req <- parseRequest . Text.unpack $ "http://localhost:9001/hello/" <> input.name
  resp <- httpBS req
  pure . decodeUtf8 $ getResponseBody resp
```
continuation of the previous suggestion; `decodeUtf8` is strict in the UTF-8 parsing sense of the word, so if we're handed invalid UTF-8 text it'll throw an exception.

if we want to use the exercise as a demonstration of error handling, one could do `either throwM pure . decodeUtf8'`.